### ISSUE TITLE
Advertise H2 via ALPN/NPN when available.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8188]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)| @Lukasa reported|
|-|-|
|Trac ID|trac#8188|
|Type|enhancement|
|Created|2016-01-29 09:58:10Z|
|Branch|https://github.com/twisted/twisted/tree/h2-alpn-8188-2|

This patch updates `twistd web` to advertise HTTP/2 via ALPN and NPN when HTTP/2 is available.

With the patch in its current form, HTTP/2 will *never* be available, but that's ok.

This is one of the sub-patches for [#7460](https://github.com/twisted/twisted/issues/7460). It'll be part of a cohesive set of patches, landed one at a time, in order to make reviewing the code substantially easier for the Twisted team, who are wildly overloaded at this time.

Currently this patch doesn't have tests: I'll work on adding those today.

Attachments:

* [8188_1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7b9/7b9a05b85592f54f37e57fe6d1cd83906c4756b4/95bb22288cbd13575780ebf5d5aa398e1782f69a.patch) (2374 bytes) - added by Lukasa on 2016-01-29 10:01:09Z - First draft patch, no tests.
* [8188_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7b9/7b9a05b85592f54f37e57fe6d1cd83906c4756b4/cb256beb262f770b84ad803bc1f380964d72cff7.patch) (3078 bytes) - added by Lukasa on 2016-01-29 14:33:17Z - Second draft patch with some testing
* [8188_3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7b9/7b9a05b85592f54f37e57fe6d1cd83906c4756b4/e42f7e4ee9191bbe2f3943240cbff7809ba400b7.patch) (2594 bytes) - added by Lukasa on 2016-04-06 12:13:44Z - Updated patch removing some unnecessary changes.

<details><summary>Searchable metadata</summary>

```
trac-id__8188 8188
type__enhancement enhancement
reporter__Lukasa Lukasa
priority__normal normal
milestone__None None
branch__h2_alpn_8188_2 h2-alpn-8188-2
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1454061490022022 1454061490022022
changetime__1474198049187875 1474198049187875
version__None None
owner__Amber_Brown__HawkOwl___hawkowl_____ Amber Brown (HawkOwl) <hawkowl@...>

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_1"></a>|@Lukasa commented|
|-|-|

This second draft patch includes some testing. I haven't yet worked out how best to test the ALPN identifiers are actually set, so I'd love a suggestion on that topic from a reviewer.

Regardless, this is now R4R.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

Thanks for the patch.

Happy to see progress on HTTP 2 support.

Adding support for twistd looks to me like one of the last patches which should be merged in the series of HTTP 2 patches.

twisted.web.http.H2_ENABLED is a new public thing and it needs documentation... but do we really need this new public API?

There is twisted.python.reflect.requireModule so you can have something like this:

```
if requireModule('twisted.web.http2') is not None:
    acceptableProtocols.insert(0, b'h2')
```

I don't know how to write a valid test which will work on any environment.

I would say that we can have 2 tests each skipped if http2 is or is not available and check how twistd web is initiated.

For the actual test, I was thinking at a high-level functional test which uses local/mocked sockets to set up the server with twistd and then a client to initiate the connection and check that H2 is advertises over HTTPS... something similar to the ALPN tests.

Leaving this for review in case someone has better comments :)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

(In [46775]) Branching to h2-alpn-8188.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @Lukasa|
|-|-|

Thank you for your contribution, Lukasa.  And thanks for taking care to split it into nice, easy-to-review chunks.

Since we don't need to land this immediately, this is all "optional" feedback except where otherwise noted, in the sense that we could probably land it mostly in the state that it's in if it were imminently necessary to close the loop on http2.  But since this is one of the last things necessary and I'm reviewing it now, please do consider addressing these in some way.

1. Mandatory point: there are new pyflakes errors.  Relatedly, I agree with adiroiban's feedback about the new "public" API of `H2_ENABLED`.  The usual idiom in Twisted for this sort of thing is:
```
#!py
try:
    from twisted.web.http2 import H2Connection
except ImportError:
    H2Connection = None
```
  which would address pyflakes's problems; I'd prefer that to a `requireModule` call.
2. Mandatory point: I also agree with adiroiban's feedback about test coverage.  The relevant thing to test is at the very least the value of the `acceptableProtocols` attribute, not the newly-minted `H2_ENABLED` flag.
3. The fact that this is in `tap.py` is somewhat unfortunate, for a few reasons.
    1. `--https=...` is sort of softly deprecated in favor of `--port ssl:...`.  Making this functionality only available when you use the "old, bad" way of doing things, with different options for different ports, is unfortunate.
    2. All Twisted HTTP servers should be able to get H2, not just those started with `twistd web`.  So putting the logic to detect whether and how to enable `acceptableProtocols` is really unfortunate.
    3. This puts the knowledge in the wrong place.  The thing that "knows" whether it should be advertising the protocol or not is `Site`.
  My recommendation for fixing this issue would be to add an optional interface that an `IProtocolFactory` can implement which the various TLS-using endpoints could check for, and honor.  This would generally solve the issue for all protocols which want to (A) use string endpoints, but also (B) want to specify application-level ALPN/NPN.  Presently, it's the constructor of the endpoint that receives the `connectionCreator` (AKA `contextFactory`), and `CertificateOptions` takes `acceptableProtocols` to its constructor.  However, bear in mind that it's the _endpoint_ which passes on its `connectionCreator` through to `TLSMemoryBIOFactory`; fraught as the OpenSSL API is here, there is actually still the opportunity to compose the `IOpenSSLServerConnectionCreator` interface, potentially manipulating the connection object on the way through, calling `set_alpn_protos` (or similar) on it before passing it on.  This could obviously be its own stand-alone ticket.

Thanks again for contributing all this work to Twisted.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_5"></a>|@Lukasa commented|
|-|-|

Thanks Glyph and Adi. I'm going to hold off on applying this review feedback until we've landed some of the others: this patch can be used to close the loop on the rest of the HTTP/2 work. However, the feedback is really appreciated and I'll definitely be taking it on board.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_6"></a>|@Lukasa removed owner|
|-|-|

Ok, I've pushed up a GitHub branch to my GitHub fork, called `8188`. This branch contains a totally different implementation of the ALPN negotiation for HTTP/2, based on using endpoints as per glyph's request. Given that, I'd like to do some brand new review.

I'm also not entirely sure at what the coverage looks like at this time, so I'd like hawkie to test that out too.

Anyway, based on that branch, ready for review!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

Please provide a link to that branch... or better, create a PR so that the basic tests are automatically triggered.

Please provide the link to the PR as a comment.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_8"></a>|@Lukasa commented|
|-|-|

Thanks adi, see https://github.com/twisted/twisted/pull/181.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Hopefully adi can complete a full review on this, but I should just say: I like what I see here.  I left a few drive-by nitpicks on the PR which may of course be ignored.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl set owner to @hawkowl<br>@hawkowl set status to `closed`|
|-|-|

In changeset 72fa6caea4e282ccf9d077195254c00e69a06d0a
```
#!CommitTicketReference repository="" revision="72fa6caea4e282ccf9d077195254c00e69a06d0a"
Merge h2-alpn-8188-2: Advertise H2 via ALPN/NPN when available.

Author: lukasa
Reviewers: adiroiban, glyph, hawkowl
Fixes: #8188
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing review from closed tickets.
